module/VuFind/src/VuFind/Captcha/AltchaFactory.php

```diff
@@ -0,0 +1,105 @@
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see
+ * <https://www.gnu.org/licenses/>.
+ *
+ * @category VuFind
+ * @package  CAPTCHA
+ * @author   Chris Hallberg <challber@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+
+namespace VuFind\Captcha;
+
+use AltchaOrg\Altcha\Hasher\Algorithm;
+use Laminas\ServiceManager\Exception\ServiceNotCreatedException;
+use Laminas\ServiceManager\Exception\ServiceNotFoundException;
+use Laminas\ServiceManager\Factory\FactoryInterface;
+use Psr\Container\ContainerExceptionInterface as ContainerException;
+use Psr\Container\ContainerInterface;
+
+/**
+ * Altcha proof-of-work CAPTCHA factory.
+ *
+ * @category VuFind
+ * @package  CAPTCHA
+ * @author   Chris Hallberg <challber@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class AltchaFactory implements FactoryInterface
+{
+    /**
+     * Create an object
+     *
+     * @param ContainerInterface $container     Service manager
+     * @param string             $requestedName Service being created
+     * @param null|array         $options       Extra options (optional)
+     *
+     * @return object
+     *
+     * @throws ServiceNotFoundException if unable to resolve the service.
+     * @throws ServiceNotCreatedException if an exception is raised when
+     * creating a service.
+     * @throws ContainerException&\Throwable if any other error occurs
+     */
+    public function __invoke(
+        ContainerInterface $container,
+        $requestedName,
+        ?array $options = null
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options passed to factory.');
+        }
+
+        $config = $container
+            ->get(\VuFind\Config\PluginManager::class)
+            ->get('config');
+
+        $secret = $config->Captcha->altcha_secret ?? null;
+
+        if (empty($secret)) {
+            throw new \Exception('Secret key needed for Altcha. See config.ini.');
+        }
+
+        $algorithm = Algorithm::from($config->Captcha->altcha_algorithm ?? 'SHA-256');
+        $maxNumber = $config->Captcha->altcha_max_number ?? 100000;
+        $saltLength = $config->Captcha->altcha_salt_len ?? 12;
+        $expiresInterval = $config->Captcha->altcha_expires_interval ?? null;
+        $params = $config->Captcha->altcha_params ?? [];
```

Comment:
...and then we need to convert all the object notation to array notation:


```suggestion
        $secret = $config['Captcha']['altcha_secret'] ?? null;

        if (empty($secret)) {
            throw new \Exception('Secret key needed for Altcha. See config.ini.');
        }

        $algorithm = Algorithm::from($config['Captcha']['altcha_algorithm'] ?? 'SHA-256');
        $maxNumber = $config['Captcha']['altcha_max_number'] ?? 100000;
        $saltLength = $config['Captcha']['altcha_salt_len'] ?? 12;
        $expiresInterval = $config['Captcha']['altcha_expires_interval'] ?? null;
        $params = $config['Captcha']['altcha_params'] ?? [];
```